### DISCUSSION TITLE
Priority Queue For User Songs

### DISCUSSION THREAD

**Discussion by Menghini** [audioling/audioling#5] [Ideas]
Both Feishin and Sonixd have an overall context queue. I can append a song to the head or tail of the queue in both of these apps. However, if I append song J, K, L, and M to the queue in either of these apps, I only have the following Situations (1 and 2):

**Situation 1 - Add Next**
Assume the queue includes (in this order): songs A, B, C, and D. If I queue a song to the the head "aka next", such as queuing (in this order) J then K then L and then M I would have the song queue M, L, K, J, A, B, C, and D.

**Situation 2 - Add Last**
Assume the queue includes (in this order): A, B, C, and D. If I queue a song to the the head "aka later", such as queuing (in this order) J then K then L and then M I would have the song queue A, B, C, D, M, L, K, J.

**The Preferable Way To Queue**
I would like to see situation 3 which has both a user queue and a context queue. Often times, I have a playlist of over 1,000 songs and I want to queue certain songs in the order of what I chose, while still having the entire playlist queued. But I am unable to do so. Consider situation 3:

**Situation 3 - User Queue**
Assume the queue already includes (in this order): songs A, B, C, and D (and many others from a random shuffle of 1,000+ songs in a playlist). If I queue a song to the user queue, such as queuing (in this order) J then K then L and then M I would then have the song queue be J, K, L, M, A, B, C, and D. It will play song J first, as it was the first one I added, then L, etc but keep songs A, B, C, and D (and possibly 1,000+ songs which were queued because I chose "shuffle" in the playlist). In other words, it would have a higher priority to queue songs manually queued by the user higher than the songs that were hit from a simple "play all" or "shuffle all"

**Comment by jeffvli** [comment]
This is basically [Spotify-style queueing](https://github.com/jeffvli/feishin/issues/640) right?

**Reply by Menghini** [reply]
Yep! Exactly!

**Reply by jeffvli** [reply]
Took me a while to understand how Spotify's queue worked, but I think I have an initial implementation working! The queue will be configurable to switch between the `Default-only` queueing style and the `Priority` style.

https://github.com/user-attachments/assets/9e9ed1b5-e53b-4e30-bf29-9cbba7f5fbfa

Edit: Not sure why the video embed isn't working, but right clicking it to save allows you to view the preview that I made.

**Reply by Menghini** [reply]
Wow. Slick. I like it! You're doing great work!

My only recommendation is also having a button to do the same thing that replicates the Spotify feature. Although I don't mind drag-and-dropping, I think a button would be great to. Although I'm not quite sure what you would use as the language between "Play Next" and "Play Last". But I know that pretty much means you have to rework how the queue is implemented.

**Reply by jeffvli** [reply]
Drag/drop isn't required for the feature.

The main `Play` button replicates Spotify's `Play` button. First track of the selection gets added to the top of the priority queue while placing the rest of the selection into the 'non-priority' or default queue.

The `Play last` button replicates Spotify's `Add to queue` button. The entire selection gets added to the end of the priority queue.

The only thing different is that I'm still keeping the `Play next` button which is also used for the default-only queue system. But when the priority queue is used, it will place the selection as follows:
1. IF the currently playing song is in the priority queue, it will be after the currently playing song
2. IF the currently playing song is in the default queue, it will be at the end of the priority queue

**Reply by Menghini** [reply]
Oh shoot. I see it now. I apologize. This is exactly what I was hoping for! Thank you!

**Comment by Menghini** [comment]
With respect to the [v0.1.0 update](https://github.com/audioling/audioling/releases/tag/audioling-v0.1.0) - I am very impressed by how easy the drag and drop is. Thank you for adding this. It goes without saying, the right click context support is a much needed feature as well for various things (add to playlist, add to queue). Obviously not a full feature release, but thought I would provide some feedback.

I know you have the ability to add a whole album/playlist to the queue, but I would also like to do that by song as well when just viewing "tracks".
